plugins/outputs/application_insights/application_insights.go

```diff
@@ -0,0 +1,410 @@
+}
+
+type ApplicationInsights struct {
+	InstrumentationKey string
+	Timeout            internal.Duration
+	DiagMsgSubscriber  DiagnosticsMessageSubscriber
+	Transmitter        TelemetryTransmitter
+	diagMsgListener    appinsights.DiagnosticsMessageListener
+}
+
+var (
+	sampleConfig = `
+## Instrumentation key of the Application Insights resource.
+instrumentationKey = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx"
+
+## Timeout on close. If not provided, will default to 5s. 0s means no timeout (not recommended).
+# timeout = "5s"
+`
+	is32Bit        bool
+	is32BitChecked bool
+)
+
+func (a *ApplicationInsights) SampleConfig() string {
+	return sampleConfig
+}
+
+func (a *ApplicationInsights) Description() string {
+	return "Send telegraf metrics to Azure Application Insights"
+}
+
+func (a *ApplicationInsights) Connect() error {
+	if a.Timeout.Duration == 0 {
+		a.Timeout = internal.Duration{Duration: time.Second * 5}
+	}
+
+	if a.Transmitter == nil {
+		a.useDefaultTransmitter()
+	}
+
+	if a.Transmitter != nil && a.DiagMsgSubscriber != nil {
+		a.diagMsgListener = a.DiagMsgSubscriber.Subscribe(func(msg string) error {
+			logOutputMsg("%s", msg)
+			return nil
+		})
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 || a.Transmitter == nil || a.InstrumentationKey == "" {
+		return nil
+	}
+
+	for _, metric := range metrics {
+		if canBeTreatedAsAggregate(metric) {
+			if telemetry := createAggregateMetricTelemetry(metric); telemetry != nil {
+				a.Transmitter.Track(telemetry)
+				continue
+			}
+		}
+
+		if telemetry := createSimpleMetricTelemetry(metric); telemetry != nil {
+			a.Transmitter.Track(telemetry)
+		}
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Close() error {
+	if a.Transmitter == nil {
+		return nil
+	}
+
+	select {
+	case <-a.Transmitter.Close():
+		logOutputMsg("Closed")
+	case <-time.After(a.Timeout.Duration):
+		logOutputMsg("Close operation timed out after %v", a.Timeout.Duration)
+	}
+
+	if a.diagMsgListener != nil {
+		// We want to listen to diagnostic messages during closing
+		// That is why we stop listening only after Close() ends (or a timeout occurs)
+		a.diagMsgListener.Remove()
+	}
+
+	return nil
+}
+
+func init() {
+	outputs.Add("application_insights", func() telegraf.Output {
+		return &ApplicationInsights{
+			DiagMsgSubscriber: appinsightsDiagnosticsMessageSubscriber{},
+		}
+	})
+}
+
+func (a *ApplicationInsights) useDefaultTransmitter() {
+	// Env variable, if present, overrides what is in the configuration
+	iKey := os.Getenv("APPINSIGHTS_INSTRUMENTATIONKEY")
+	if iKey != "" {
+		a.InstrumentationKey = iKey
+	}
+
+	if a.InstrumentationKey != "" {
+		a.Transmitter = NewAppinsightsTransmitter(a.InstrumentationKey)
+	}
+}
+
+func canBeTreatedAsAggregate(metric telegraf.Metric) bool {
+	// If we can identify the Sum and Count properties, then we are good
+	hasSum := metric.HasField("value") || metric.HasField("sum")
```

Comment:
We hardcoded the rule of how an metric can be treated as aggregated, which would be hard to follow. What if the field name is "sample" or "sampling" instead of "samples", or "total" instead of "sum"?
Someone will have to decide whether a metric is aggreated, but I hope it's done by the input plugin, e.g., AppMetric input plugin will know the property name that means aggregated plugin, or some override plugin with more flexibility.
As an output plugin, we can just rely on this property Metric.IsAggregate.

Reply:
telegraf.Metric.IsAggregate() is just a flag on the (Telegraf) metric type. From Application Insights perspective, to create an AppInsights aggregate metric, we have to know at least the sum of all values and the number of samples, so telegraf.Metric.IsAggregate() alone does not help us with that task. 

Application Insights aggregate metric is a concept specific to AppInsighst, so it *has to be the AppInsights output plugin* that makes the decision whether a Telegraf metric can be interpreted as an AppInsights aggregate metric or not. We might eventually need more flexibility here in terms of how this interpretation is made, but it quickly gets complicated and I do not want to try to solve this in the first iteration.